File: arduino_mcuota.py
#!/usr/bin/python -u
###
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
#
# Copyright 2015 Arduino Srl (http://www.arduino.org/)
#
# authors:
# sergio@arduino.org
#
###

import argparse, requests, sys, os

_version = '0.0.1'
_serv_file = 'otafile'

def _split_file(filename, nlines):
	"""Splits the input file every nlines lines."""
	counter = 1
	file_list = []
	input_file = open(filename, 'r').read().split('\n')

	for lines in range(0, len(input_file), nlines):
		# First, get the list slice
		output_data = input_file[lines:lines + nlines]

		output_file = open(filename + "." + str(counter) + '.out', 'w')
		output_file.write('\n'.join(output_data))
		output_file.close()
		#print "Chunk #" + str(counter) + ": " + output_file.name
		file_list.append(output_file.name)
		counter += 1

	return file_list

def _upload_files(file_list, filename, address, port = 80):
	try:
		url = "http://" + address + ":" + str(port) + "/"+ _serv_file
		totchunk = len(file_list)
		numchunk = 0
		payload = {'totchunk': str(totchunk), 'numchunk': str(numchunk) }
		headers = {'Content-type': 'multipart/form-data'}
		print "Sending " + filename + " to host " + address
		for index, fl in enumerate(file_list):
		#for fl in iter(file_list):
			index += 1
			f = open(fl, 'rb')

			files = {'file': f}
			payload['numchunk'] = str(index)
			r = requests.post(url, files=files, params=payload, headers=headers, timeout=15 )
			print "["+str(index) + " / " + str(len(file_list)) + "] Done..."

			#print str(r.status_code) + " " + r.content
		print 'Upload ' + r.content
	except requests.RequestException as req_error:
		print req_error
		sys.exit(-1)

def _remove_files(file_list):
	try:
		for fl in iter(file_list):
			#print "Deleting file : " + fl
			os.remove(fl)
	except Exception as ex:
		print ex
		sys.exit(-1)

parser = argparse.ArgumentParser(prog='arduino_mcuota',
	#usage='%(prog)s ',
	prefix_chars='-',
	formatter_class=argparse.ArgumentDefaultsHelpFormatter,
	description='Arduino MCU-OTA Upload Tool - Upload sketches to the microcontroller of the board Over The Air via ESP8266 - Copyright 2016 Arduino Srl',
	epilog='')

requiredNamed = parser.add_argument_group('required named arguments')
requiredNamed.add_argument('-f', '--file', type=file, help='firmware file to upload', required=True)
requiredNamed.add_argument('-i', '--ip', help='ip of the esp8266 based board to upload to', required=True)

parser.add_argument('-p', '--port', type=int , default=80, help='network port')
parser.add_argument('-l', '--lines', type=int , default=20, help='max lines for each splitted files')
#parser.add_argument('-v', '--version', dest='lines', help='version number')

args = parser.parse_args()

_list = _split_file(args.file.name, args.lines)
_upload_files(_list, args.file.name, args.ip, args.port)
_remove_files(_list)

#sys.exit(0)

File: arduino_mcuota.go
package main

import (
    "os"
    "fmt"
    "strconv"
    "io"
    "bufio"
    "math"
    "gopkg.in/resty.v0"
    "flag"
)

var filename string
var address string
var port int
var nlines int

const endpoint = "otafile"


func upload_files(list []string){
    url_base := "http://" + address + ":" + strconv.Itoa(port) + "/" + endpoint
    totchunk := len(list)
    
    fmt.Println("Sending " + filename + " to host " + address)
    
    for index, item := range list{        
    
        url := url_base + "?totchunk=" + strconv.Itoa(totchunk) + "&numchunk=" + strconv.Itoa(index+1)
        resp, err := resty.R().
        SetFile("files", item).
        Post(url)
        
        if resp.StatusCode() == 200 || err != nil {
            fmt.Println("["+strconv.Itoa(index+1)+" / "+strconv.Itoa(totchunk)+"] Done")
                
            if index+1==totchunk {
                fmt.Println("Upload all done")
            }
        } else
        if resp.StatusCode() > 399 || err == nil {
            fmt.Println("["+strconv.Itoa(index+1)+" / "+strconv.Itoa(totchunk)+"]")
            fmt.Print("Error: ")
            fmt.Println(err)
        }
    }
    clean(list)
}

func split(){
    file_list := []string {}

    lines := nlines
     file, err := os.Open(filename)
     if err != nil {
             panic(err)
     }

     defer file.Close()
     
     index := 1
     counter := 0
     var output string
        reader := bufio.NewReader(file)
        for {
            line, _, err := reader.ReadLine()
            
            if err == io.EOF {
                 break
            }

          output += string(line) + "\n"
         
            if math.Mod(float64(index) , float64(lines)) == 0 || index == getFileLines(filename) {
                     output_file, err := os.Create(filename + "." + strconv.Itoa(counter+1) + ".out")
                     output += string(line) + "\n"
                     if err != nil {
                            fmt.Println(err)
                        }
                     output_file.WriteString(output)
                     output_file.Close();
                     file_list = append (file_list, filename + "." + strconv.Itoa(counter+1) + ".out")
                     output = ""
                     counter ++   
            }
         index ++ 
     }    
    upload_files(file_list)
}

func clean(list []string){
    counter:=1
    for _, item := range list{
        os.Remove(item)
        counter++
    }
}

func get_args(){
    flag.StringVar(&filename, "f", "", "firmware file to upload (required)")
    flag.StringVar(&address, "i", "", "ip of the esp8266 based board to upload to (required)")
    flag.IntVar(&port, "p", 80, "network port")
    flag.IntVar(&nlines, "l", 20, "max lines for each splitted files")

    flag.Parse()
    
    if len(filename)>0 && len(address)>0 {
        split()
    } else {
        fmt.Print("usage : ")
        fmt.Println("arduino_mcuota [-h] -f FILE -i IP [-p PORT] [-l LINES]")
        fmt.Print("error: ")
        
        if len(filename) == 0 {
            fmt.Println("-f is required")
        }
        if len(address) == 0 {
            fmt.Println("-i is required")
        }
    }
}

func getFileLines(filename string)(int){
        tot:=0
        file, err := os.Open(filename)
        if err != nil {
             panic(err)
        }
        defer file.Close()

        reader := bufio.NewReader(file)
        for {
                 _, _, err := reader.ReadLine()
                if err == io.EOF {
                         break
                 }
                tot++
        }
        return tot
}


func main(){
    get_args()
}